### PR TITLE
(feat: M2-7478 & M2-7479) Idle Timer implementation

### PR DESCRIPTION
**Pull request by moiskillnadne**
### 📝 Description
🔗 [Jira Ticket M2-7478](https://mindlogger.atlassian.net/browse/M2-7478)
🔗 [Jira Ticket M2-7479](https://mindlogger.atlassian.net/browse/M2-7479)

Changes include:

- useIdleTimer hook
- InactivityTracker component

### CODE REVIEW

File: src/widgets/ProtectedRoute/InactivityTracker.tsx [hunk 1]
```diff
@@ -1,50 +1,34 @@
...
 
     return () => {
-      Object.values(events).forEach((item) => {
-        window.removeEventListener(item, onActivityEventHandler);
+      events.forEach((item) => {
+        window.removeEventListener(item, listener);
       });
     };
-  }, [onActivityEventHandler]);
+  }, [IdleTimer]);
```
Comment (BamMironov) [inline src/widgets/ProtectedRoute/InactivityTracker.tsx:31]: Please double-check to ensure everything is working the way it should. `IdleTimer` is recreated on every render by `useIdleTimer`. Hence, the effect and its clean function run every render.

File: src/widgets/ProtectedRoute/InactivityTracker.tsx [hunk 2]
```diff
@@ -1,50 +1,34 @@
...
+    const listener = () => IdleTimer.start('InactivityTracker');
+
+    events.forEach((item) => {
+      window.addEventListener(item, listener);
     });
 
     return () => {
-      Object.values(events).forEach((item) => {
-        window.removeEventListener(item, onActivityEventHandler);
+      events.forEach((item) => {
```
Comment (BamMironov) [inline src/widgets/ProtectedRoute/InactivityTracker.tsx:27]: The events getting added/removed in two places: `InactivityTracker.tsx` and `useIdleTimer.tsx`. Please stick with one place only.
I'd recommend passing the `events` array as a parameter to the `useIdleTimer` and then managing them inside `useIdleTimer`. Also, `IdleTimer` should have an additional function - `stop` that will take over the cleaning up so taht you could invoke it inside the effect's clean up function. Example:

```
...
const IdleTimer = useIdleTimer({
  time: {
    hours: 0,
    minutes: 15,
  },
  events,
  onFinish: logout,
});
...
useEffect(() => {
  IdleTimer.start('InactivityTracker');

  return () => {
    IdleTimer.stop('InactivityTracker');
  }
}, [IdleTimer]);
```

File: src/widgets/Survey/ui/PassingScreen.tsx [hunk 3]
```diff
@@ -277,6 +288,39 @@ const PassingScreen = () => {
     onTimerEnd: hasNextStep ? onNext : openSubmitModal,
   });
 
+  const IdleTimer = useIdleTimer({
+    time: context.event.timers.idleTimer, // Value should be provided from event DTO, if null - timer will not work
```
Comment (BamMironov) [inline src/widgets/Survey/ui/PassingScreen.tsx:292]: The comment is nice for understanding the use case. However, the same thing can be accomplished by writing the code. You can either move `time` to the `start` function and explicitly check if `time` is not null (as it was mentioned earlier) OR explicitly has a flag, for instance "enabled":
```
const IdleTimer = useIdleTimer({
   time: context.event.timers.idleTimer,
   enabled: !!context.event.timers.idleTimer
   onFinish: () => {
      if (props.onTimerFinish) {
        props.onTimerFinish();
      }
    },
});
```

File: src/widgets/Survey/ui/PassingScreen.tsx [hunk 4]
```diff
@@ -277,6 +288,39 @@ const PassingScreen = () => {
...
+    const listener = () => IdleTimer.start('SurveyIdleTracker');
+
+    events.forEach((item) => {
+      window.addEventListener(item, listener);
+    });
+
+    return () => {
+      events.forEach((item) => {
+        window.removeEventListener(item, listener);
+      });
```
Comment (BamMironov) [inline src/widgets/Survey/ui/PassingScreen.tsx:320]: Same thing about the events array

File: src/widgets/Survey/ui/ScreenManager.tsx [hunk 5]
```diff
@@ -54,9 +58,26 @@ export const ScreenManager = ({ openTimesUpModal }: Props) => {
...
+    }
+
+    if (
+      autoCompletionState.activityIdsToSubmit.length ===
+      autoCompletionState.successfullySubmittedActivityIds.length
+    ) {
+      return;
+    }
+
+    if (openTimesUpModal) {
```
Comment (BamMironov) [inline src/widgets/Survey/ui/ScreenManager.tsx:76]: Ain't no way `useEffect` should do something based on the component's props. `useEffect` is all about synchronization with the external system like Browser API (e.g.setTimeout):
https://react.dev/reference/react/useEffect
https://react.dev/learn/synchronizing-with-effects

Therefore, `openTimesUpModal` should be invoked in a function that is a result of a certain action

**Review by BamMironov — CHANGES_REQUESTED** [review]
Please consider changes below:

File: src/features/IdleTimer/lib/useIdleTimer.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { useCallback } from 'react';
+
+import { getMsFromHours, getMsFromMinutes, HourMinute } from '~/shared/utils';
+import useTimer from '~/shared/utils/useTimer';
+
+// Events that we consider as activity
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/features/IdleTimer/lib/useIdleTimer.ts:6]: Done

File: src/features/IdleTimer/index.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+export * from './lib/useIdleTimer';
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/features/IdleTimer/index.ts:1]: Done

File: src/features/IdleTimer/lib/useIdleTimer.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  time: HourMinute | null;
+  onFinish: () => void;
+};
+
+export const useIdleTimer = (props: Props) => {
+  const { resetTimer, setTimer } = useTimer();
+
+  const onTimerExpire = useCallback(() => {
+    // Listener clean up. Removes the existing event listener from the window
+    Object.values(events).forEach((item) => {
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/features/IdleTimer/lib/useIdleTimer.ts:19]: Done

File: src/features/IdleTimer/lib/useIdleTimer.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    // Execute onComplete function
+    if (props.onFinish) {
+      props.onFinish();
+    }
+  }, [props, resetTimer]);
+
+  const start = useCallback(
+    (idleTimerName?: string) => {
+      if (!props.time) {
```
[the remark this answers is not in the document]
Comment (moiskillnadne) [reply src/features/IdleTimer/lib/useIdleTimer.ts:31]: Done

File: src/widgets/ProtectedRoute/InactivityTracker.tsx (hunk 1 above)
Comment (moiskillnadne) [reply src/widgets/ProtectedRoute/InactivityTracker.tsx:31]: Done

File: src/widgets/Survey/ui/PassingScreen.tsx (hunk 4 above)
Comment (moiskillnadne) [reply src/widgets/Survey/ui/PassingScreen.tsx:320]: Done

File: src/widgets/Survey/ui/PassingScreen.tsx (hunk 3 above)
Comment (moiskillnadne) [reply src/widgets/Survey/ui/PassingScreen.tsx:292]: Yes, good suggestion. Done

File: src/widgets/Survey/ui/ScreenManager.tsx (hunk 5 above)
Comment (moiskillnadne) [reply src/widgets/Survey/ui/ScreenManager.tsx:76]: Done

File: src/shared/constants/events.ts [hunk 10]
```diff
@@ -0,0 +1,2 @@
+// Events that we consider an activity
+export const events = ['click', 'keypress', 'keydown', 'mousemove', 'mousedown', 'touchstart'];
```
Comment (BamMironov) [inline src/shared/constants/events.ts:2]: Let's call it `interactionEvents`

File: src/widgets/Survey/ui/PassingScreen.tsx [hunk 11]
```diff
@@ -1,14 +1,14 @@
 import { useCallback, useContext, useEffect, useMemo } from 'react';
 
 import { AutoCompletionModel } from '../../../features/AutoCompletion';
+import { events } from '../../../shared/constants';
```
Comment (BamMironov) [inline src/widgets/Survey/ui/PassingScreen.tsx:4]: Please import from ~/shared/constants

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/shared/constants/events.ts (hunk 10 above)
Comment (moiskillnadne) [reply src/shared/constants/events.ts:2]: Done

File: src/widgets/Survey/ui/PassingScreen.tsx (hunk 11 above)
Comment (moiskillnadne) [reply src/widgets/Survey/ui/PassingScreen.tsx]: Done

**Review by BamMironov — APPROVED** [review]

**Pull request merged**